openmdao/visualization/meta_model_viewer/meta_model_visualization.py

```diff
@@ -0,0 +1,898 @@
+                             fill_alpha=alphas.tolist())
+
+        # Set the right_plot data source to new values
+        self.bottom_plot_scatter_source.data = dict(
+            bot_slice_x=x_data,
+            bot_slice_y=np.repeat(y_value, self.resolution))
+
+        self.contour_plot.line(
+            'bot_slice_x', 'bot_slice_y', source=self.bottom_plot_scatter_source, color='black',
+            line_width=2)
+
+        return self.bot_plot_fig
+
+    def _unstructured_training_points(self, compute_distance=False, source='bottom'):
+        """
+        Calculate the training points and returns and array containing the position and alpha.
+
+        Parameters
+        ----------
+        compute_distance : bool
+            If true, compute the distance of training points from surrogate line.
+        source : str
+            Which subplot the method is being called from.
+
+        Returns
+        -------
+        array
+            The array of training points and their alpha opacity with respect to the surrogate line
+        """
+        # Input training data and output training data
+        x_training = self.model_ref._training_input
+        training_output = np.squeeze(stack_outputs(self.model_ref._training_output), axis=1)
+
+        # Index of input/output variables
+        x_index = self.x_input_select.options.index(self.x_input_select.value)
+        y_index = self.y_input_select.options.index(self.y_input_select.value)
+        output_variable = self.output_names.index(self.output_select.value)
+
+        # Vertically stack the x/y inputs and then transpose them
+        infos = np.vstack((x_training[:, x_index], x_training[:, y_index])).transpose()
+        if not compute_distance:
+            return infos
+
+        points = x_training.copy()
+
+        # Normalize so each dimension spans [0, 1]
+        points = np.divide(points, self.limit_range)
+        dist_limit = np.linalg.norm(self.dist_range * self.limit_range)
+        scaled_x0 = np.divide(self.input_point_list, self.limit_range)
+
+        # Query the nearest neighbors tree for the closest points to the scaled x0 array
+        # Nearest points to x slice
+        if x_training.shape[1] < 3:
+
+            tree = cKDTree(points)
+            # Query the nearest neighbors tree for the closest points to the scaled x0 array
+            dists, idx = tree.query(
+                scaled_x0, k=len(x_training), distance_upper_bound=self.dist_range)
+
+            # kdtree query always returns requested k even if there are not enough valid points
+            idx_finite = np.where(np.isfinite(dists))
+            dists = dists[idx_finite]
+            idx = idx[idx_finite]
+
+        else:
+            dists, idx = self._multidimension_input(scaled_x0, points, source=source)
+
+        # data contains:
+        # [x_value, y_value, ND-distance, func_value, alpha]
+
+        data = np.zeros((len(idx), 4))
+        for dist_index, j in enumerate(idx):
```

Comment:
I wouldn't use 'idx' to represent a collection of indices since 'idx' is singular, not plural.  Maybe 'idxs' instead?